clippy_lints/src/boxed_void.rs

```diff
@@ -0,0 +1,50 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::path_def_id;
+use clippy_utils::ty::is_c_void;
+use rustc_hir::{Expr, ExprKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::RawPtr;
+use rustc_middle::ty::TypeAndMut;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks if we're passing a `c_void` raw pointer to `Box::from_raw(_)`
+    ///
+    /// ### Why is this bad?
+    /// This is a known pitfall of FFI, particularly with some C or C++ APIs.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::ffi::c_void;
+    /// let ptr = Box::into_raw(Box::new(42usize)) as *mut c_void;
+    /// let _ = unsafe { Box::from_raw(ptr) };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::ffi::c_void;
+    /// let ptr = Box::into_raw(Box::new(42usize)) as *mut c_void;
+    /// let _ = unsafe { Box::from_raw(ptr as *mut usize) };
+    /// ```
+    #[clippy::version = "1.66.0"]
+    pub BOXED_VOID,
+    suspicious,
+    "don't create a Box from a raw void pointer"
+}
+declare_lint_pass!(BoxedVoid => [BOXED_VOID]);
+
+impl LateLintPass<'_> for BoxedVoid {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let ExprKind::Call(box_from_raw, [arg]) = expr.kind
+        && let ExprKind::Path(QPath::TypeRelative(ty, seg)) = box_from_raw.kind
+        && seg.ident.name == sym!(from_raw)
+        && path_def_id(cx, ty).map_or(false, |id| Some(id) == cx.tcx.lang_items().owned_box()) {
+            let arg_kind = cx.typeck_results().expr_ty(arg).kind();
+            if let RawPtr(TypeAndMut { ty, .. }) = arg_kind {
+                if is_c_void(cx, *ty) {
+                span_lint(cx, BOXED_VOID, expr.span, "don't create a Box from a raw void pointer");
```

Comment:
I would word the lint message a bit different: ``"creating a `Box` from a raw void pointer"`` and then add a suggestion (with `span_lint_and_sugg`) to tell the user to cast the pointer first.

Reply:
We cannot know what type to cast to in general. So the suggestion could at best contain a placeholder, which might be confusing. Wouldn't some help text be better suited here?

Reply:
We could suggest `as *mut _` with applicability `HasPlaceholder`. But I agree that this might be confusing. So I'm also fine with just a help message.

Reply:
Updated into this

```rust
span_lint_and_help(cx, RAW_BOX_FROM_VOID_PTR, expr.span, "creating a `Box` from a raw void pointer", Some(arg.span), "cast this to the actual type");
```

Reply:
I would write `cast this to a pointer of the actual type`.